### PR TITLE
🏗🚀 Rewrite and speed up `gulp build` using `esbuild` and `@babel/core`

### PR DESCRIPTION
**Pull request by rsimha**
This is another in a series of PRs that modernize our development tasks.

**PR highlights:**
- Rewrite `gulp build` without file streaming
- Replace `browserify ` with `esbuild` and `@babel/core` for a considerable increase in speed
- Remove `gulp-regexp-sourcemaps`, `gulp-rename`, `gulp-sourcemaps`, `vinyl-source-stream`, and `watchify` 

**Results:** This brings down the time taken for a [full unminified build](https://app.circleci.com/pipelines/github/ampproject/amphtml/2058/workflows/cf6015b6-a8cf-48f3-8277-4ee497ef5567/jobs/21181/parallel-runs/0/steps/0-107) of AMP from ~5m to ~1m.

**References:**
- https://esbuild.github.io/api/
- https://babeljs.io/docs/en/babel-core

Partial fix for #32585

### CODE REVIEW

**Comment by kristoferbaxter** [comment]
Added @samouri to review since he was independently implementing the same thing.

File: build-system/tasks/helpers.js
Comment (samouri) [inline build-system/tasks/helpers.js:405]: Ideally we can extract the wrapper + the internalRuntimeVersion this babel plugin.
For the wrapper, we can take advantage of `banner` and `footer`, e.g.

```js
  const sentinel = '<%= contents %>';
  const start = wrapper.indexOf(sentinel);
  const header = wrapper.slice(0, start);
  const footer = wrapper.slice(start + sentinel.length);
```

File: build-system/tasks/helpers.js
Comment (samouri) [inline build-system/tasks/helpers.js:443]: I believe we can take advantage of both `watch` and `incremental` options

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js]: Super idea. Added a function for this.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:459]: The reason I don't think we should do this is that for extensions, the watcher needs to first build all CSS and then invoke the chain of `buildExtensionJs()` -> `compileUnminifiedJs()` -> `esbuild.build()`. I'm concerned that adding `watch` and `incremental` here will short-circuit the chain.

https://github.com/ampproject/amphtml/blob/11b1fcae9e6c7c7d829234e7a01f94033ea4dc72/build-system/tasks/extension-helpers.js#L435-L446

Okay to punt on this until I can get tests to work correctly, and attempt optimizations in a follow up PR?

**Comment by rsimha** [comment]
With excellent and timely help from @samouri, we now have a clean workflow with `esbuild` + `babel`. Unminified build [now takes ~1m](https://app.circleci.com/pipelines/github/ampproject/amphtml/2053/workflows/8792b468-b058-4c39-8c07-7c4bd8868d4a/jobs/21093) (down from ~5m) and all tests are [green](https://app.circleci.com/pipelines/github/ampproject/amphtml/2053/workflows/8792b468-b058-4c39-8c07-7c4bd8868d4a/jobs/21103).

@kristoferbaxter @jridgewell Heads up that one of you is going to have to provide owners approval after this PR goes through a full review.

File: build-system/tasks/helpers.js
Comment (samouri) [inline build-system/tasks/helpers.js:462]: Can we add `$internalRuntimeVersion$` to the define set?

File: build-system/tasks/helpers.js
Comment (samouri) [inline build-system/tasks/helpers.js:441]: nit: s/processWrapper/splitWrapper

Comment (rsimha) [reply build-system/tasks/helpers.js]: Done.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:462]: Tried doing this in a few different ways, none seemed to work. I think the double quotes around `$internalRuntimeVersion$` are messing with `esbuild`. Escaping the double quotes didn't work either. Will punt on this for now and see if there are other tricks to use during the `gulp dist` rewrite.

Comment (samouri) [reply build-system/tasks/helpers.js:462]: I don't believe you need the quotation marks

File: build-system/tasks/helpers.js
Comment (samouri) [inline build-system/tasks/helpers.js:426]: how are we cache busting?

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:462]: Sorry, I edited my previous comment with more info. Doing this without quotes results in no substitution.

The original code is:

https://github.com/ampproject/amphtml/blob/ee68dd7be3707618d7c077d493e9ad3778a97a84/src/internal-version.js#L26-L28

And `esbuild` converts that code to:

image

Perhaps `define` only works for constant identifiers, and not string literals?

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js]: Oooh, excellent point. We weren't busting the cache earlier, but I've added code that does so and tested it as follows:

```sh
# Clean build state
gulp clean

# Launch lazy-build dev server
gulp

# Load http://127.0.0.1:8000/examples/article.amp.html
# Prints `Powered by AMP ⚡ HTML – Version `

# Modify `src/amp.js` and change the line to `Powered by AMP2`
# `amp.js` is rebuilt

# Reoad http://127.0.0.1:8000/examples/article.amp.html
# Prints `Powered by AMP2 ⚡ HTML – Version `
```

Incidentally, this is the exact same mechanism used by `pre-closure-babel.js` used by `gulp dist`. I'll revisit the caching there when I refactor `gulp dist` in case we need to do the same cache busting there too.

File: build-system/tasks/helpers.js
Comment (samouri) [reply build-system/tasks/helpers.js:464]: Darn. I think you're right. Can punt to a future cleanup PR but do you think it'd be reasonable to convert `internalRuntimeVersion` into a constant?

**Review by samouri — APPROVED** [review]

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:464]: Good idea, it might be possible to make it a const, but I think that's best left for a follow up PR where the change can be tested in isolation. Will do soon.

File: build-system/tasks/helpers.js
Comment (samouri) [reply build-system/tasks/helpers.js]: Sounds great. Just tested it locally and it worked!

File: build-system/tasks/helpers.js
Comment (kristoferbaxter) [inline build-system/tasks/helpers.js:133]: Nit: Given the proximity here, it could be easy for someone to confuse `watch` and `options.watch`.

Could we continue to rename the export from `chokidar`? `{watch: fileSystemWatch}` or similar?

File: build-system/tasks/helpers.js
Comment (kristoferbaxter) [inline build-system/tasks/helpers.js:345]: This a really subtle but clean change in the PR. Nice!

File: build-system/tasks/helpers.js
Comment (kristoferbaxter) [inline build-system/tasks/helpers.js:419]: Was there any logic here regarding sourcemaps? Do we need to ensure they remain accurate with the swap of versions value?

File: build-system/tasks/helpers.js
Comment (kristoferbaxter) [inline build-system/tasks/helpers.js:426]: An interesting experiment in the future would be to use the async helper to execute these steps in parallel unless I am not reading this code correctly.

File: build-system/tasks/helpers.js
Comment (kristoferbaxter) [reply build-system/tasks/helpers.js:464]: +1 for a followup, also worth checking if this is an outstanding feature request to `esbuild`.

**Comment by samouri** [comment]
thinking out loud: would a much more efficient cache-bust be to use have the chokidar watcher remove the filepath from the map?

**Review by kristoferbaxter — APPROVED** [review]
So many nice changes here, thank you for driving this thru.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js]: Fair point. Done.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:426]: When I added logging for `build.onLoad` and ran a full `gulp build`, it looked like `esbuild` was bundling all extensions in parallel. Will experiment some more in future.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:419]: I'm pretty sure we don't need any more special logic. `esbuild` is bundling code that was already babel-transformed, and generating a sourcemap for the final result. When I loaded `dist/amp.js` and `dist/amp.js.map` in http://sokra.github.io/source-map-visualization, the symbols seemed to line up.

**Comment by rsimha** [comment]
> thinking out loud: would a much more efficient cache-bust be to use have the chokidar watcher remove the filepath from the map?

The purpose of this `babel` cache is to prevent O(N^2) transforms from occurring when O(N) entry points are bundled (e.g. `log.js` is used by almost all entry points, but we only want to transform it once unless the code inside it has changed). The chokidar watcher acts at a higher level (file level for runtime JS entry points, and directory level for extensions), and triggers a full rebuild of the JS entry point / extension when a change is detected. So I'm not sure that it can be used in place of the inner babel cache.

With the babel caching (and cache busting) in place here, subsequent rebuilds of JS entry points and extensions will be faster, as will the initial full-build of everything.

File: build-system/tasks/helpers.js
Comment (kristoferbaxter) [reply build-system/tasks/helpers.js:419]: Perfect, thanks for the confirmation!

File: build-system/tasks/helpers.js
Comment (jridgewell) [inline build-system/tasks/helpers.js:405]: Why isn't this just done with a babel plugin?

File: build-system/tasks/helpers.js
Comment (jridgewell) [reply build-system/tasks/helpers.js:419]: Only the line with `$internalRuntimeVersion$` is going to be incorrect, and it still needs to be handled. Moving this to an actual Babel plugin will resolve it.

File: build-system/tasks/helpers.js
Comment (jridgewell) [inline build-system/tasks/helpers.js:424]: When is `transform` not defined? And when do we fall into the below path?

File: build-system/tasks/helpers.js
Comment (jridgewell) [reply build-system/tasks/helpers.js:419]: I'm actually really confused why any sourcemap works here? If we're running a babel-transform, then we need to return a map from Babel to es-build. That's the only way it could correctly resolve the mappings through the transformed code.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:424]: I added `transform` because I used [this babel plugin](https://github.com/nativew/esbuild-plugin-babel/blob/main/src/index.js) as a starting point.

I added logging and it turns out that `transform` is never defined. From reading https://esbuild.github.io/plugins/, I think it's safe to remove the `if (transform)` block. Will do in a follow up.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:405]: Because I was trying to minimize changes in this PR 😃 

I'm working on a follow up that tries to make `$internalRuntimeVersion$` a const variable instead of a literal string. If that's a bad idea due to how closure works, then making it a babel plugin SGTM.

Comment (samouri) [reply build-system/tasks/helpers.js:405]: I said similarly in a different comment. addressing in a followup

File: build-system/babel-config/unminified-config.js
Comment (rsimha) [reply build-system/babel-config/unminified-config.js:48]: This was needed because of how we were previously adding the wrapper to individual files before bundling. I think I can get rid of this in a follow up PR.

File: build-system/tasks/helpers.js
Comment (samouri) [reply build-system/tasks/helpers.js:424]: He likely took inspo from here: https://github.com/nativew/esbuild-plugin-babel/blob/main/src/index.js

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:405]: Turns out we can't make this a `const` because we'd then have to re-jig how tests run. Right now, we conveniently assume that the string is not replaced for tests. Changing this to a `const` yields this new failure:
```sh
  ReferenceError: INTERNAL_RUNTIME_VERSION is not defined
      at internalRuntimeVersion (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/src/internal-version.js:27:3)
      at getRtvVersion (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/src/mode.js:135:15)
      at getMode_ (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/src/mode.js:85:18)
      at getMode (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/src/mode.js:54:28)
      at new Log (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/src/log.js:155:16)
      at dev (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/src/log.js:834:22)
      at initLogConstructor (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/src/log.js:757:3)
      at Object.require.702.../build/ampdoc.css (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/src/runtime.js:51:1)
      at o (/var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/node_modules/browser-pack/_prelude.js:1:1)
      at /var/folders/nv/ycwrjg_j32gdz6s542xlxy0m0037zm/T/node_modules/browser-pack/_prelude.js:1:1

```

Can I leave it to one of y'all to move this into a babel plugin?

File: build-system/babel-config/unminified-config.js
Comment (rsimha) [reply build-system/babel-config/unminified-config.js:48]: Removing this in #32696.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:424]: Removing this in #32696.

File: build-system/tasks/helpers.js
Comment (rsimha) [reply build-system/tasks/helpers.js:464]: Changing this to a const is more complicated than initially assumed due to https://github.com/ampproject/amphtml/pull/32663#discussion_r577210915. Moving it to a babel plugin seems to be the best option.

File: build-system/tasks/helpers.js
Comment (jridgewell) [reply build-system/tasks/helpers.js:419]: Resolving with https://github.com/ampproject/amphtml/pull/32698.

**Comment by samouri** [comment]
> So I'm not sure that it can be used in place of the inner babel cache.

The watcher knows each file that is modified. The babel cache is on a file by file basis. When the watcher detects that a file has been modified, we can remove it from the cache.

This allows us to **do less**. We would no longer need to perform a `sha` of each files contents when it wasn't modified

**Comment by rsimha** [comment]
> The watcher knows each file that is modified. The babel cache is on a file by file basis. When the watcher detects that a file has been modified, we can remove it from the cache.

Thanks for pushing back on this. You make a good point here. My initial concern was that for extensions, we watch an entire directory and therefore may not know whether the change was to a CSS file or JS file or something else. However, I see that you've managed to capture this info in #32744 via `modifiedFile`. I retract my previous statement and will review that PR 😃

**Pull request merged**
